### DISCUSSION TITLE
Problems sending rpc command via code

### DISCUSSION THREAD

**Discussion by KevinAStein** [robotology/community#116] [Q&A] [answered]
Hi,

I wanted to send commands to open **rpc ports** using C++.  I followed the client example from http://www.yarp.it/rpc_ports.html, but I don't manage to get the same result as when using `yarp rpc` from within a terminal.

Thanks,
Kevin

**Comment by towardthesea** [comment]
Hi @KevinAStein, could you please provide more details on how you did it and the response, e.g terminal snapshot. Without the details, I am afraid that nobody can give you advice!

Regards,
Phuong

**Comment by KevinAStein** [comment]
Hi,

in terminal I use `yarp rpc /portname/rpc:i` and then send my command to the module. 
Instead, I want to use something like:

``` cpp
  RpcClient port;
  port.open("/app/Port:o");
  Network::connect("/app/Port:o","/portname/rpc:i");
  Bottle cmd;
  cmd.addString("do something");
  Bottle response;
  if (port.write(cmd, response)) {
    cout << response.toString() << endl;
  }
```

If I use the help command I get the same output. But using other commands it seems that the code version is not waiting for the module to finish the task and instantly responds in failure.

**Comment by towardthesea** [comment]
Hi @KevinAStein ,

From the above mentioned code, I only can say that your created `client` will open an `rpc port` and then connect to the the `/app/Port:o server port`. Then, it sends a `do something` string to the `client port` and displays the response from the `server`. If it is your expected result, you got it. 
Actually I am afraid I cannot say more without knowing what did you do with your `server`!

P/S: I assume that you followed the  tutorial at http://www.yarp.it/rpc_ports.html and got the same results!

Regards,
Phuong

**Comment by KevinAStein** [comment]
Yes, I understand what my code is supposed to do, but why is there a difference between running it manually in the command line interface or within C++? Am I missing something?

**Comment by pattacini** [comment]
Hi @KevinAStein 

Would you mind to post your code, both the client and the server part, using [**`Gist`**](https://gist.github.com)? This way, we'll be able to compile and run it and see what's going on.

**Comment by KevinAStein** [comment]
I am trying to interact with `WholeBodyDynamicsTree` for the icub robot. I want to send the calibration command when initializing my module instead of doing it manually in a terminal every time. There is only the lines posted above.

**Comment by pattacini** [comment]
Then, it might be a sort of mismatch between the sent command's format (a string?) and the expected one.

In `ARE` we reset `WBD` using this [piece of code](https://github.com/robotology/icub-main/blob/master/src/modules/actionsRenderingEngine/include/iCub/MotorThread.h#L267-L279) and it's working fine (disregard the call to `Time::delay` since it has nothing to do with the rpc comm).

However, I'm not into the `WBDTree` rpc format; I'd suggest you to check that out, therefore.

**Comment by pattacini** [comment, marked as answer]
One pitfall could be also that

``` cpp
cmd.addString("do something");
```

is different from

``` cpp
cmd.addString("do");
cmd.addString("something");
```

Don't know if this applies to your case though.

**Comment by KevinAStein** [comment]
Thanks, that actually solved my problem!
